GH-113464: Add the JIT to What's New (#133486)
python/cpython: Doc/whatsnew/3.14.rst

(Contributed by Łukasz Langa in :gh:`131507`.)

.. _whatsnew314-jit-compiler:

> An experimental just-in-time compiler
> -------------------------------------

Reviewer: I think this is a clearer headline, especially as 3.13 included the same item. Weakly-held, though.
```suggestion
Binary releases for the experimental just-in-time compiler
----------------------------------------------------------
```

Official macOS and Windows release binaries now include an *experimental* JIT
compiler. Although it is **not** recommended for production use, it can be
enabled by setting the :envvar:`PYTHON_JIT` environment variable. Downstream
source builds can be configured with :option:`--enable-experimental-jit=yes-off`
for similar behavior.

This is still early, foundational work; the performance impact of enabling the
JIT typically ranges from 10% slower to 20% faster, depending on workload. The

> most significant missing functionality at this time is the inability for native
> debuggers and profilers like ``gdb`` and ``perf``  to unwind through JIT frames

Reviewer 2: ```suggestion
most significant missing functionality at this time is that native
debuggers and profilers like ``gdb`` and ``perf`` are unable to unwind through JIT frames
```
Reads better to me but feel free to disregard

> Official macOS and Windows release binaries now include an *experimental* JIT
> compiler. Although it is **not** recommended for production use, it can be
> enabled by setting the :envvar:`PYTHON_JIT` environment variable. Downstream
> source builds can be configured with :option:`--enable-experimental-jit=yes-off`
> for similar behavior.
> 
> This is still early, foundational work; the performance impact of enabling the
> JIT typically ranges from 10% slower to 20% faster, depending on workload. The
> most significant missing functionality at this time is the inability for native
> debuggers and profilers like ``gdb`` and ``perf``  to unwind through JIT frames
> (Python debuggers and profilers, like :mod:`pdb` or :mod:`profile`, work without
> modification). Free-threaded builds do not support JIT compilation.
> 
> :func:`sys._jit.is_available` can be used to determine if your Python executable
> supports JIT compilation. To tell if JIT compilation has been enabled for the
> current process, use :func:`sys._jit.is_enabled`.
> 
> Please report any bugs or major performance regressions that you encounter!

Reviewer: Quick suggestion through the GH web UI, not checked for spelling or line length. Feel free to edit at will.
```suggestion
The official macOS and Windows release binaries now include an *experimental*
just-in-time (JIT) compiler. Although it is **not** recommended for production
use, it can be tested by setting :envvar:`PYTHON_JIT=1 <PYTHON_JIT>` as an
environment variable. Downstream source builds and redistributors can use the
:option:`--enable-experimental-jit=yes-off` configuration option for similar behavior.

The JIT is at an early stage and still in active development. As such, the
typical performance impact of enabling the JIT can range from 10% slower
to 20% faster, depending on workload. To aid in testing and evaluation,
a set of introspection functions has been provided in the :data:`sys._jit`
namespace. :func:`sys._jit.is_available` can be used to determine if the executable
has been built with the JIT enabled, and :func:`~sys._jit.is_enabled` records if JIT
compilation is currently enabled.

Currently, the most significant missing feature is the inability for native debuggers
and profilers like ``gdb`` or ``perf``  to unwind through JIT frames.
However, Python debuggers and profilers, such as :mod:`pdb` or :mod:`profile`,
work without modification. Free-threaded builds do not support JIT compilation.

Please report any bugs or major performance regressions that you encounter!
```
